### PR TITLE
Prevent main from having a where clause.

### LINKED ISSUE #50714: Linker error when main has a false trait bound
This programme gives a linker error

```rust
fn main() where fn(&()): Eq {}
```

Compiler output:

<details>

```text
   Compiling playground v0.0.1 (file:///playground)
error: linking with `cc` failed: exit code: 1
  |
  = note: "cc" "-Wl,--as-needed" "-Wl,-z,noexecstack" "-m64" "-L" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "/playground/target/debug/deps/playground-0e82fec49f1d853f.playground0.rcgu.o" "-o" "/playground/target/debug/deps/playground-0e82fec49f1d853f" "/playground/target/debug/deps/playground-0e82fec49f1d853f.crate.allocator.rcgu.o" "-Wl,--gc-sections" "-pie" "-Wl,-z,relro,-z,now" "-nodefaultlibs" "-L" "/playground/target/debug/deps" "-L" "/playground/target/debug/build/backtrace-sys-abbe17c29cebc5cc/out/.libs" "-L" "/playground/target/debug/build/miniz-sys-7a0db4796c5e7c10/out" "-L" "/usr/lib/x86_64-linux-gnu" "-L" "/playground/target/debug/build/ring-67ac7987243a7fd2/out" "-L" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib" "-Wl,--start-group" "-Wl,-Bstatic" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd-912509c5d8c60ccd.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libpanic_unwind-008c42aad548e7dd.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_jemalloc-f68d8e8245e507c6.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libunwind-0fe53371ec419e32.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc_system-9dec1cbd51097ce1.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liblibc-6a4fb915dd86d140.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/liballoc-7ebba6af2d3cc324.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libstd_unicode-7a26f8b3cf380464.rlib" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcore-b0c2d164a9741309.rlib" "-Wl,--end-group" "/root/.rustup/toolchains/stable-x86_64-unknown-linux-gnu/lib/rustlib/x86_64-unknown-linux-gnu/lib/libcompiler_builtins-66d072e25a9acee3.rlib" "-Wl,-Bdynamic" "-l" "dl" "-l" "rt" "-l" "pthread" "-l" "pthread" "-l" "gcc_s" "-l" "c" "-l" "m" "-l" "rt" "-l" "pthread" "-l" "util" "-l" "util"
  = note: /usr/lib/gcc/x86_64-linux-gnu/5/../../../x86_64-linux-gnu/Scrt1.o: In function `_start':
          (.text+0x20): undefined reference to `main'
          collect2: error: ld returned 1 exit status
          

error: aborting due to previous error

error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

</details>

cc #48214 - which will make this issue easier to come across.

### PR DESCRIPTION
**Pull request by matthewjasper**
Closes #50714

Should this have a crater run?

cc #48557, #48214

r? @nikomatsakis

### CODE REVIEW

**Comment by varkor** [comment]
You also need to add this check for `start` functions:
https://github.com/rust-lang/rust/blob/27acb9b3c033e1dd74cc72310a5a346527d3e704/src/librustc_typeck/lib.rs#L251-L258

**Comment by nikomatsakis** [comment]
So, this cannot affect libraries, so the potential impact is pretty limited. I suspect we can get by without a crater run.

@bors dele

**Comment by nikomatsakis** [comment]
r=me once we check `start` functions as well like @varkor suggested

@bors delegate=matthewjasper

File: src/librustc_typeck/diagnostics.rs [hunk 1]
```diff
@@ -4519,6 +4519,28 @@ impl Foo for () {
...
+}
+```
+"##,
+
+E0647: r##"
+It is not possible to define `start` with a where clause.
+Erroneous code example:
+
+```compile_fail,E0646
+fn start() where i32: Copy { // error: start function is not allowed to have
```
Comment (varkor) [inline src/librustc_typeck/diagnostics.rs]: Your tests look correct, but this isn't a correct example of a start function.

**Pull request merged**